File: 1046-last-stone-weight/1046-last-stone-weight.py
import heapq

class Solution:
    def lastStoneWeight(self, stones):
        # Convert into max heap by using negative values
        stones = [-stone for stone in stones]
        heapq.heapify(stones)

        # Continue until one or no stone remains
        while len(stones) > 1:
            y = -heapq.heappop(stones)  # largest
            x = -heapq.heappop(stones)  # second largest

            # If stones are not equal
            if y != x:
                heapq.heappush(stones, -(y - x))

        # Return remaining stone or 0
        return -stones[0] if stones else 0

File: 1046-last-stone-weight/1046-last-stone-weight.java
import java.util.PriorityQueue;

class Solution {
    public int lastStoneWeight(int[] stones) {

        // Max Heap
        PriorityQueue<Integer> pq = new PriorityQueue<>((a, b) -> b - a);

        // Add all stones into heap
        for (int stone : stones) {
            pq.add(stone);
        }

        // Smash stones until one or none left
        while (pq.size() > 1) {

            int y = pq.poll(); // largest
            int x = pq.poll(); // second largest

            // If not equal, push difference
            if (y != x) {
                pq.add(y - x);
            }
        }

        // Return remaining stone or 0
        return pq.isEmpty() ? 0 : pq.poll();
    }
}